### PR TITLE
fix for when first half pin found, sometimes M5 not received before NACK

### PR DESCRIPTION
**Pull request by feitoi**
After first half pin was found, and when M5 message not received before NACK message then considers as UNKNOWN_ERROR and re-trying last pin.

see #327

Fix only for second half pin

### CODE REVIEW

**Comment by feitoi** [comment]
When I went to see in the history why the variable `pin_count` of crack.c is a float type and not an int type, I just found out that there was something about FAKE NACK.
I copied some code from there and tested it on the router and saw that when that NACK problem happens, the code was "MESSAGE_TIMEOUT" (0x0010).
The pin is 12425082.
see the output for firf half pin:
```
root@kali:~# reaver -i wlan1mon -b 74:EA:3A:**:**:** -c 11 -vvNL -p 12425082

Reaver v1.6.6-git-42-g44ea045 WiFi Protected Setup Attack Tool
Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>

[+] Switching wlan1mon to channel 11
[?] Restore previous session for 74:EA:3A:**:**:**? [n/Y] n
[+] Waiting for beacon from 74:EA:3A:**:**:**
[+] Received beacon from 74:EA:3A:**:**:**
[+] Vendor: AtherosC
[+] Trying pin "12425082"
[+] Sending authentication request
[+] Sending association request
[+] Associated with 74:EA:3A:**:**:** (ESSID: AP-Test)
[+] Sending EAPOL START request
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received M1 message
[+] Sending M2 message
[+] Received M3 message
[+] Sending M4 message
[+] Received WSC NACK (reason: 0x0010)
[!] WARNING: Potential FAKE NACK!
[+] Sending WSC NACK
[!] WPS transaction failed (code: 0x04), re-trying last pin
[+] Trying pin "12425082"
[+] Sending authentication request
[+] Sending association request
[+] Associated with 74:EA:3A:**:**:** (ESSID: AP-Test)
[+] Sending EAPOL START request
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received M1 message
[+] Sending M2 message
[+] Received M3 message
[+] Sending M4 message
[+] Received M5 message
[+] Sending M6 message
[+] Received M7 message
[+] Sending WSC NACK
[+] Sending WSC NACK
[+] Pin cracked in 26 seconds
[+] WPS PIN: '12425082'
[+] WPA PSK: 'password'
[+] AP SSID: 'AP-Test'
```
see the output for second half pin:
```
root@kali:~# reaver -i wlan1mon -b 74:EA:3A:**:**:** -c 11 -vvNL -p 1242

Reaver v1.6.6-git-41-g9d6aac5 WiFi Protected Setup Attack Tool
Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>

[+] Switching wlan1mon to channel 11
[?] Restore previous session for 74:EA:3A:**:**:**? [n/Y] 
[+] Restored previous session
[+] Waiting for beacon from 74:EA:3A:**:**:**
[+] Received beacon from 74:EA:3A:**:**:**
[+] Vendor: AtherosC
[+] Trying pin "12425679"
[+] Sending authentication request
[+] Sending association request
[+] Associated with 74:EA:3A:**:**:** (ESSID: AP-Test)
[+] Sending EAPOL START request
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received M1 message
[+] Sending M2 message
[+] Received M3 message
[+] Sending M4 message
[+] Received M5 message
[+] Sending M6 message
[+] Received WSC NACK (reason: 0x1003)
[+] Sending WSC NACK
[+] Trying pin "12420001"
[+] Sending authentication request
[+] Sending association request
[+] Associated with 74:EA:3A:**:**:** (ESSID: AP-Test)
[+] Sending EAPOL START request
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received identity request
[+] Sending identity response
[+] Received M1 message
[+] Sending M2 message
[+] Received M3 message
[+] Sending M4 message
[+] Received WSC NACK (reason: 0x0010)
[!] WARNING: Potential FAKE NACK!
[+] Sending WSC NACK
[!] WPS transaction failed (code: 0x04), re-trying last pin
^C
```

**Comment by rofl0r** [comment]
> about FAKE NACK.
> I copied some code from there and tested it on the router and saw that when that NACK problem happens, the code was "MESSAGE_TIMEOUT" (0x0010).

interesting. you say that when we get MESSAGE_TIMEOUT as NACK reason, we should retry rather than considering that as a pin error, right ?

**Comment by feitoi** [comment]
> we should retry rather than considering that as a pin error, right?

Yes.

I only tested to a router because I only know it for now. So the nake_code is 0x0010.

I discovered this router unintentionally because I inserted a check to issue a "BELL" when receive NACK without receiving an M5 to indicate possible pin changed. 

I made a collection of data, all possible nack_codes after M3 or M5 are:
0x0012 (majority)
0x1003 (majority)
0x000F (when starts WPS lock)
0x0000

**Comment by rofl0r** [comment]
> I made a collection of data, all possible nack_codes after M3 or M5 are:
> 0x0012 (majority)
> 0x1003 (majority)
> 0x000F (when starts WPS lock)
> 0x0000

processing these status code could make the brute-force attempt much more robust.
do we have a description of these error codes somewhere in source (like macro names or comments) ?

**Comment by feitoi** [comment]
> do we have a description of these error codes somewhere in source (like macro names or comments) ?

Yes, they are in the def.h file, just 0x1003 is not yet. `enum nack_code{...}`

**Comment by feitoi** [comment]
In Wireshark, the description of 0x1003 is: Configuration Error: Unknown (0x1003)

File: src/exchange.c [hunk 1]
```diff
@@ -231,6 +231,27 @@ enum wps_result do_wps_exchange()
 			set_timeout_is_nack(0);
 
 			ret_val = KEY_REJECTED;
+
+			/* Check the reason code for the received NACK message */
+			if (!get_pin_string_mode()) {
```
Comment (rofl0r) [inline src/exchange.c:236]: i'd think we could just unconditionally check for get_nack_reason() == MESSAGE_TIMEOUT here and set ret_val to UNKNOWN_ERROR (assuming that means "try again")

File: README.md [hunk 2]
```diff
@@ -91,6 +91,7 @@ Advanced Options:
 	-p, --pin=<wps pin>             Use the specified pin (may be arbitrary string or 4/8 digit WPS pin)
 	-d, --delay=<seconds>           Set the delay between pin attempts [1]
 	-l, --lock-delay=<seconds>      Set the time to wait if the AP locks WPS pin attempts [60]
+	-I, --ignore-reason             Ignore checking of reason code for NACK [False]
```
Comment (rofl0r) [inline README.md:94]: i'd rather just always do the right thing than adding yet another command line option.

File: src/exchange.c [hunk 3]
```diff
@@ -472,6 +493,11 @@ enum wps_type process_wps_message(const void *data, size_t data_size)
                                         case MESSAGE_TYPE:
                                                 type = (uint8_t) element_data[0];
                                                 break;
+					case CONFIGURATION_ERROR:
+						/* Check element_data length */
+						if (element.length == 2)
+							set_nack_reason(htons(*( (uint16_t*) element_data)));
```
Comment (rofl0r) [inline src/exchange.c:499]: i guess WPA_GET_BE16(element_data) would be more readable (and no potential issues with unaligned access).
also maybe we should initialize the nack reason to something like -1 before entering this block, so we dont mistakenly interpret a previous nack as one.
does a nack always cause the CONFIGURATION_ERROR condition here to get triggered ?

Comment (feitoi) [reply src/exchange.c:499]: > does a nack always cause the CONFIGURATION_ERROR condition here to get triggered ?

Yes, the CONFIGURATION_ERROR and MESSAGE_TYPE (NACK(0x0E)) are in the same packet, and its are processed.

**Comment by rofl0r** [comment]
cool, thanks. have you been able to verify whether the new code works as expected ?

**Review by Shahzadeh3am — APPROVED** [review]

File: src/exchange.c [hunk 4]
```diff
@@ -231,6 +231,17 @@ enum wps_result do_wps_exchange()
...
 
 			ret_val = KEY_REJECTED;
+
+			/* Check the reason code for the received NACK message */
+			if (get_nack_reason() == MESSAGE_TIMEOUT) {
+				ret_val = UNKNOWN_ERROR;
+				cprintf(WARNING, "[!] WARNING: Potential FAKE NACK!\n");
+			}
+			/* Got NACK instead of an M5 message, when cracking second half */
+			else if (last_msg == M3 && get_key_status() == KEY2_WIP) {
```
Comment (feitoi) [inline src/exchange.c:241]: By logic, yet missing the checking of `get_pin_string_mode()`

**Comment by feitoi** [comment]
> have you been able to verify whether the new code works as expected ?

Yes, it works on the router 74:EA:3A:**:**:** and other routers with a valid PIN.
I don't have a router with "illegitimate" WPS PIN, so I don't know how it will be.

File: src/exchange.c (hunk 4 above)
Comment (rofl0r) [reply src/exchange.c]: sorry, could you elaborate what you mean ?

Comment (feitoi) [reply src/exchange.c]: When we specified a not valid pin or empty pin, the first half pin and second half pin are irrelevants and consequently the status KEY2_WIP is invalid/fake after received M5, for to reinforce not to enter the block it is necessary `get_pin_string_mode()`

Comment (rofl0r) [reply src/exchange.c]: i see. can you push a commit addressing that ?

Comment (feitoi) [reply src/exchange.c]: I force-pushed the last commit 387678d with that

**Comment by kimocoder** [comment]
May you tag a new release or version to it, and we'll push to Kali 👍  and thanks!

**Comment by rofl0r** [comment]
> May you tag a new release or version to it, and we'll push to Kali

hmmmm... i understand your enthusiasm but i'm reluctant to make a release whenever we fix a minor issue (especially if it's not widely tested yet). what i'd propose instead is that you use the last commit hash instead of a tag: `https://github.com/USER/REPO/archive/SHA1.tar.gz` and label it reaver-1.6.6-$hash or similar.

**Comment by kimocoder** [comment]
Better. Thanks 🥇

**Comment by feitoi** [comment]
> I made a collection of data, all possible nack_codes after M3 or M5 are:
0x0012 (majority)
0x1003 (majority)
0x000F (when starts WPS lock)
0x0000

see above https://github.com/t6x/reaver-wps-fork-t6x/pull/329#issuecomment-841671500 and next 2

**Comment by feitoi** [comment]
I found 2 more nack_code after M3:
0x1200 (only 1 AP so far)
0xFFFF (only 1 AP so far)

**Comment by rofl0r** [comment]
well my question is: why don't we use UNKNOWN_CFG_ERROR ?

**Comment by feitoi** [comment]
We haven't found a use for it yet.
Until this commit, `enum nack_code` is not even used.
I added it to make the list of nack_code more complete. Maybe the name didn't come out very well. Better would be **UNKNOWN_CONFIG_ERROR**.

**Comment by rofl0r** [comment]
> Maybe the name didn't come out very well. Better would be UNKNOWN_CONFIG_ERROR.

don't worry, as long as its unused it's ok as is.

**Pull request merged**